# Exercise 4.5 - Data Consistency

## This script contains:
#### 01. Importing libraries
#### 02. Importing data
#### 03. Data consistency checks
#### 04. Exploring products data
#### 05. Task step 2
#### 06. Task step 3
#### 07. Task step 4
#### 08. Task step 5
#### 09. Task step 6
#### 10. Task step 7
#### 11. Task step 9

# 01. Importing Libraries

In [1]:
# import libraries

import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [2]:
# creating path to Instacart folder

path = r'/Users/eastongorman/Library/CloudStorage/OneDrive-Personal/Data Analytics/06-2023 Instacart Basket Analysis'

In [3]:
# creating products dataframe

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# creating orders dataframe

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 03. Data Consistency Checks

In [5]:
# create a dataframe

df_test = pd.DataFrame()

In [6]:
# create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# changing data type

df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# finding missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# creating subset of products dataframe containing only null values

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# (rows, colums)

df_prods.shape

(49693, 5)

In [14]:
# creating subset of products dataframe with no missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# check if rows decreased

df_prods_clean.shape

(49677, 5)

In [16]:
# check for full duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
#creating new subset from df_prods_clean subset that contains no duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
# check if rows decreased

df_prods_clean_no_dups.shape

(49672, 5)

# 04. Exporting products data

In [20]:
# exporting df_prods_clean_no_dups as products_checked.csv

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 05. Task step 2

In [21]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

## No anomolies found in the orders dataframe.

In [22]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [23]:
# finding item with max price

df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] == 99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

## The max for prices is 99,999 which is most likely a typo. Further investigation shows that 2% milk has the wrong price.

In [24]:
# checkinhg if any other item is over $50

df_prods_clean_no_dups[df_prods_clean_no_dups['prices']> 50.0]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

## The decimal place for both the cottage cheese and milk need to be moved 4 spaces left.

In [25]:
# change the price of the cottage cheese to  1.49000 and the price of the milk to 9.99

df_prods_clean_no_dups = df_prods_clean_no_dups.replace ({"prices":{99999.0: 9.99, 14900.0:1.49 }})

In [26]:
# check that changes have been made

df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"] == 21553]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

       prices  
21554    1.49

In [27]:
# check that changes have been made

df_prods_clean_no_dups[df_prods_clean_no_dups["product_id"]==33664]

product_id           product_name  aisle_id  department_id  prices
33666       33664  2 % Reduced Fat  Milk        84             16    9.99

# 06. Task step 3

In [28]:
# Check for mixed-type data in df_ords

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# 07. Task step 4
## If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

## No columns were printed, so there are no columns with mixed-type data in df_ords.

# 08. Task Step 5

In [29]:
# check for missing values in df_ords

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [30]:
# creating subset of orders dataframe containing only null values

df_ords_nan = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [31]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

## All of the missing values in the days_since_last_order column have an order_number of 1. This indicates the first time a customer placed an order, so it's not possible to have a last order.

## The max for user_id and the number of missing values is both 206,209, or there's one missing value for each customer.

# 09. Task step 6
## Address the missing values using an appropriate method.

## I've chosen to do nothing with the missing values. Deleting the missing values is a bad idea because it would also delete all the data from customers who have only placed one order, and imputing the value using the mean or median is also a bad idea because the imputed value wouldn't reflect a new customer. 

## A value of 1 in the user_id column can be used as a flag for a new customer.

# 10. Task step 7

In [32]:
# check for duplicate values in df_ords

df_ords_dups = df_ords[df_ords.duplicated()]

In [33]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

## There are no duplicates in df_ords.

# 11. Task step 9
## products data exported above in section 04.

In [34]:
# exporting df_ords as orders_checked.csv

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))